Add support for account recovery

oxtoacart opened the pull request:
- [x] Do the tests pass? Consistently?
- [x] Did this change improve test coverage?
- [ ] Has it been reviewed/approved by relevant teammates?
- [x] Can it be QA’d in staging or something like staging?
- [ ] Is the code in question being linted? If not, consider adding a linter step to CI. If yes, make sure the linter is happy.
- [x] Do we know how we’re going to deploy this?
- [x] Are we clear on what the support and maintenance impact is?
- [ ] Did you create new documentation? Is it accessible and in the right place?
- [ ] Has existing documentation been updated?
- [ ] Have you logged tickets for related technical debt with the label “techdebt”?

oxtoacart committed:
Added basic recovery capability

oxtoacart committed:
Now deriving identity key from recovery key

oxtoacart committed:
Downgraded to older Kotlin

oxtoacart commented on build.gradle:
```diff
@@ -1,7 +1,6 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 buildscript {
     ext.kotlin_version = "1.4.32"
-    ext.ktor_version = "1.5.2"
```
Drive-by, unused

oxtoacart committed:
Code review cleanup

oxtoacart committed:
Allow retrying from client workers

oxtoacart committed:
Revert "Allow retrying from client workers"

This reverts commit 53cab441cefea0c965adb1d326acb5b96820a1bb.

oxtoacart committed:
Improved error handling when unable to find chat number by identity key

oxtoacart committed:
Revert "Improved error handling when unable to find chat number by identity key"

This reverts commit d40213133736f48cab8020be73b68a86f8c79e7c.

soltzen commented on messaging/src/androidTest/java/io/lantern/messaging/MessagingTest.kt:
```diff
@@ -2117,6 +2123,56 @@ class MessagingTest : BaseMessagingTest() {
         }
     }
 
+    @Test
```
NIT it took me a while to understand the flow. A small explanation of what's actually being setup, executed and asserted would go a long way

soltzen replied:
I took the liberty of doing this. See my commit and elaborate, pls

soltzen commented on messaging/src/main/java/io/lantern/messaging/Messaging.kt:
```diff
@@ -241,6 +273,39 @@ class Messaging(
+        cryptoWorker.close()
+        authenticatedClientWorker.disconnect()
+        db.clear()
+        // TODO: use constant-time implementation of base32 decoding
```
Duplicate: already mentioned above

soltzen committed:
[gitignore] Nuke .idea

soltzen committed:
[readme] Add testing section

soltzen committed:
fixup! Added basic recovery capability

soltzen commented on messaging/src/androidTest/java/io/lantern/messaging/MessagingTest.kt:
```diff
@@ -2117,6 +2123,56 @@ class MessagingTest : BaseMessagingTest() {
+                newDB.use { cat1DB ->
+                    newDB.use { cat2DB ->
+                        newMessaging(dogDB, "dog").with { dog ->
+                            newMessaging(cat1DB, "cat").with { cat1 ->
```
NIT

```suggestion
                            newMessaging(cat1DB, "cat1").with { cat1 ->
```

and change the other to cat2, just for the reader to be sure this name doesn't affect the outcome

soltzen commented on messaging/src/androidTest/java/io/lantern/messaging/MessagingTest.kt:
```diff
@@ -2117,6 +2123,56 @@ class MessagingTest : BaseMessagingTest() {
+                                        "hi dog"
+                                    )
+
+                                    cat1.recover(cat2.recoveryCode)
```
I'm a bit confused: how is this working? `cat2` has no relation to `cat1`. Why is `cat2`'s recoveryKey recovering `cat1`?

Also, changing this to `cat1.recover(cat1.recoveryCode)` would yield the same result, right?

soltzen commented on messaging/src/main/java/io/lantern/messaging/Messaging.kt:
```diff
@@ -241,6 +273,39 @@ class Messaging(
+    /**
+     * Recovers the messaging system using the given recoveryKey. All existing data will be erased.
+     */
+    fun recover(recoveryCode: String) {
```
Exit early if recoveryCode is empty

soltzen commented on messaging/src/androidTest/java/io/lantern/messaging/MessagingTest.kt:
```diff
@@ -2117,6 +2123,56 @@ class MessagingTest : BaseMessagingTest() {
 
+    @Test
+    fun testRecovery() {
+        testInCoroutine {
```
This only tests the happy path. Would be great to also test:
- when a recovery key does not work
- when a recovery key is empty

soltzen reviewed (changes requested):
Looks great and clean @oxtoacart.

Apart from the inline comments, I'm still having trouble understanding the full system. It would be fantastic to document how account recovery works in the README as we discussed it here: https://wdynhnkxvsdx.slack.com/archives/C0K2M2FCN/p1636593614096200. This'll make it easier for us to improve the feature a couple of months from now.

oxtoacart committed:
Update messaging/src/androidTest/java/io/lantern/messaging/MessagingTest.kt

Co-authored-by: Alec Soltzen <83119415+soltzen@users.noreply.github.com>

oxtoacart commented:
> It would be fantastic to document how account recovery works in the README as we discussed it here:

Great point! We have a ticket to start creating proper documentation that can be consumed by users and external developers, as it's important for our legitimacy to explain how the system works as a whole. Now that we're nearing feature-complete, I'll have time to start on that and I'll include a detailed discussion of account recovery in that documentation.

oxtoacart commented:
@soltzen Thanks for your review! In particular, thanks for catching the missing validation on recovery key, I guess that's what I get for only testing the happy path :)

I've added validation there and updated the test to test the exceptional path, plus hopefully clarified a couple of things in the test too.

oxtoacart commented:
@soltzen Since I had already gone on to make additional changes on top of these in another [PR](https://github.com/getlantern/messaging-android/pull/45), I fixed the issues you mentioned in a [commit there](https://github.com/getlantern/messaging-android/commit/626f877555579e377b5656c8f0b59bd931e8e323).

shockron22 commented on messaging/src/main/java/io/lantern/messaging/RecoveryKey.kt:
```diff
@@ -0,0 +1,54 @@
+ */
+@Throws(InvalidKeyException::class)
+internal fun RecoveryKey.keyPair(label: String): ECKeyPair {
+    val derived = derive(label, 32)
```
My only thought here is to make sure to use a string of sufficient length or you don't gain as much by using HKDF. Same for below method. Perhaps you could have some form of master seed or secret as part of the key label?

"An important property of HKDF is that it does not amplify entropy but does allow a large source of weaker entropy to be utilized more evenly and effectively."

oxtoacart commented:
@soltzen @shockron22 Thank you both for the review!

@shockron22 I'm going to go ahead and merge, but please let me know if you're still concerned about the amount of entropy in the RecoveryKey and we can address in a separate PR.
